File: app/src/main/java/crux/bphc/cms/fragments/ForumFragment.java
package crux.bphc.cms.fragments;


import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.squareup.picasso.Picasso;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.List;
import java.util.Locale;

import crux.bphc.cms.R;
import crux.bphc.cms.helper.CourseRequestHandler;
import crux.bphc.cms.helper.CourseRequestHandler.CallBack;
import crux.bphc.cms.interfaces.ClickListener;
import crux.bphc.cms.models.forum.Discussion;
import crux.bphc.cms.widgets.HtmlTextView;
import io.realm.Realm;
import io.realm.RealmResults;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link ForumFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class ForumFragment extends Fragment {

    public static final String FORUM_ID_KEY = "forum_id";
    public static final String COURSE_NAME_KEY = "courseName";

    private int FORUM_ID = 1;
    private String COURSE_NAME;

    private ForumAdapter mAdapter;
    private Realm realm;

    private SwipeRefreshLayout mSwipeRefreshLayout;
    private TextView mEmptyView;

    private CourseRequestHandler courseRequestHandler;

    private Context context;

    public ForumFragment() {
        // Required empty public constructor
    }

    public static ForumFragment newInstance(int forumId, String courseName) {
        ForumFragment fragment = new ForumFragment();
        Bundle args = new Bundle();
        args.putInt(FORUM_ID_KEY, forumId);
        args.putString(COURSE_NAME_KEY, courseName);
        fragment.setArguments(args);
        return fragment;
    }

    public static ForumFragment newInstance() {
        return newInstance(1, "Site News");
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            FORUM_ID = getArguments().getInt(FORUM_ID_KEY);
            COURSE_NAME = getArguments().getString(COURSE_NAME_KEY);
        }
        realm = Realm.getDefaultInstance();
        courseRequestHandler = new CourseRequestHandler(this.getActivity());
        context = getContext();
    }

    @Override
    public void onStart() {
        if(getActivity() != null) {
            getActivity().setTitle(COURSE_NAME);
        }
        super.onStart();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        realm = Realm.getDefaultInstance();
        return inflater.inflate(R.layout.fragment_forum, container, false);
    }

    @Override
    public void onViewCreated(@NotNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mEmptyView = view.findViewById(R.id.tv_empty);

        mSwipeRefreshLayout = view.findViewById(R.id.swipeRefreshLayout);
        mSwipeRefreshLayout.setOnRefreshListener(this::makeRequest);

        ClickListener mClickListener = (object, position) -> {
            Discussion discussion = (Discussion) object;
            DiscussionFragment fragment = DiscussionFragment.newInstance(discussion.getId(), COURSE_NAME);
            FragmentTransaction transaction = requireActivity()
                    .getSupportFragmentManager()
                    .beginTransaction();

            transaction.addToBackStack(null);
            transaction.replace(((ViewGroup) requireView().getParent()).getId(), fragment, "ForumDetail");
            transaction.commit();

            return true;
        };

        RecyclerView mRecyclerView = view.findViewById(R.id.site_news);
        final LinearLayoutManager layoutManager = new LinearLayoutManager(context);
        mRecyclerView.setLayoutManager(layoutManager);

        mAdapter = new ForumAdapter(mClickListener, new ArrayList<>());
        mRecyclerView.setAdapter(mAdapter);

        makeRequest();
    }

    private void makeRequest() {
        mSwipeRefreshLayout.setRefreshing(true);
        courseRequestHandler.getForumDiscussions(FORUM_ID, new CallBack<List<Discussion>>()
        {
            @Override
            public void onResponse(List<Discussion> discussions) {
                if (discussions.size() == 0) {
                    mSwipeRefreshLayout.setRefreshing(false);
                    mEmptyView.setVisibility(View.VISIBLE);
                } else {
                    for (Discussion discussion : discussions) discussion.setForumId(FORUM_ID); // TODO replace with lambda
                    mEmptyView.setVisibility(View.GONE);
                    mAdapter.clearDiscussions();
                    mAdapter.addDiscussions(discussions);

                    // reset cached data
                    final RealmResults<Discussion> results = realm.where(Discussion.class).equalTo("forumId", FORUM_ID).findAll();
                    realm.executeTransaction(realm -> {
                        results.deleteAllFromRealm();
                        realm.copyToRealm(mAdapter.getDiscussions());
                    });
                    mSwipeRefreshLayout.setRefreshing(false);
                }
            }

            @Override
            public void onFailure(String message, Throwable t){
                mSwipeRefreshLayout.setRefreshing(false);

                // Load and display cached discussion data from database, if present
                RealmResults<Discussion> results = realm.where(Discussion.class).equalTo("forumId", FORUM_ID).findAll();
                List<Discussion> dbDiscussions = realm.copyFromRealm(results);
                if (dbDiscussions.size() == 0) {
                    mEmptyView.setVisibility(View.VISIBLE);
                    Toast.makeText(context, "No cached data. Check connection and refresh", Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(context, "Loading cached data. Refresh for latest", Toast.LENGTH_SHORT).show();
                    mAdapter.clearDiscussions();
                    mAdapter.addDiscussions(dbDiscussions);
                }
            }
        });
    }

    public static String formatDate(int seconds) {
        GregorianCalendar cal = new GregorianCalendar();
        cal.setTimeInMillis((long) seconds * 1000);
        int day = cal.get(Calendar.DAY_OF_MONTH);
        String month = cal.getDisplayName(
                Calendar.MONTH,
                Calendar.SHORT,
                Locale.getDefault());
        int year = cal.get(Calendar.YEAR);
        return day + " " + month + ", " + year;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        realm.close();
    }

    private static class ForumAdapter extends RecyclerView.Adapter<ForumAdapter.ForumViewHolder> {

        private final List<Discussion> mDiscussions;
        private final ClickListener mClickListener;

        public ForumAdapter(ClickListener clickListener, List<Discussion> discussions) {
            mClickListener = clickListener;
            mDiscussions = discussions;
        }

        public void addDiscussions(List<Discussion> discussions) {
            mDiscussions.addAll(discussions);
            notifyDataSetChanged();
        }

        public List<Discussion> getDiscussions() {
            return mDiscussions;
        }

        public void clearDiscussions() {
            mDiscussions.clear();
            notifyDataSetChanged();
        }

        @NotNull
        @Override
        public ForumViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            return new ForumViewHolder(inflater.inflate(R.layout.row_forum, parent, false));
        }

        @Override
        public void onBindViewHolder(ForumViewHolder holder, int position) {
            Discussion discussion = mDiscussions.get(position);
            holder.setIsRecyclable(false);
            holder.bind(discussion);
        }

        @Override
        public int getItemCount() {
            return mDiscussions.size();
        }


        public class ForumViewHolder extends RecyclerView.ViewHolder {

            private final ImageView mUserPic;
            private final ImageView mPinned;
            private final TextView mSubject;
            private final TextView mUserName;
            private final TextView mModifiedTime;
            private final HtmlTextView mMessage;

            public ForumViewHolder(View itemView) {
                super(itemView);
                itemView.setOnClickListener(v -> {
                    int position = getLayoutPosition();
                    mClickListener.onClick(mDiscussions.get(position), position);
                });

                mUserPic = itemView.findViewById(R.id.user_pic);
                mSubject = itemView.findViewById(R.id.subject);
                mUserName = itemView.findViewById(R.id.user_name);
                mModifiedTime = itemView.findViewById(R.id.modified_time);
                mMessage = itemView.findViewById(R.id.message);
                mPinned = itemView.findViewById(R.id.pinned);

                itemView.findViewById(R.id.click_wrapper).setOnClickListener(v -> {
                    int position = getLayoutPosition();
                    mClickListener.onClick(mDiscussions.get(position), position);
                });

            }

            public void bind(Discussion discussion) {
                Picasso.get().load(discussion.getUserPictureUrl()).into(mUserPic);
                mSubject.setText(discussion.getSubject());
                mUserName.setText(discussion.getUserFullName());
                mModifiedTime.setText(formatDate(discussion.getTimeModified()));
                mMessage.setText(HtmlTextView.parseHtml((discussion.getMessage())));
                if(!discussion.isPinned()){
                    mPinned.setVisibility(View.GONE);
                }
            }
        }
    }
}

File: app/src/main/java/crux/bphc/cms/fragments/ForumFragment.kt
package crux.bphc.cms.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.commit
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import crux.bphc.cms.R
import crux.bphc.cms.app.Urls
import crux.bphc.cms.databinding.FragmentForumBinding
import crux.bphc.cms.databinding.RowForumBinding
import crux.bphc.cms.helper.CourseDataHandler
import crux.bphc.cms.helper.CourseRequestHandler
import crux.bphc.cms.interfaces.ClickListener
import crux.bphc.cms.models.forum.Discussion
import crux.bphc.cms.utils.Utils
import io.realm.Realm
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.IOException
import java.util.*

/**
 * A simple [Fragment] subclass.
 * Use the [ForumFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class ForumFragment : Fragment() {

    private var courseId = -1
    private var forumId = 1
    private var courseName: String = ""

    private lateinit var courseRequestHandler: CourseRequestHandler
    private lateinit var binding: FragmentForumBinding

    private lateinit var mAdapter: Adapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = FragmentForumBinding.inflate(layoutInflater)

        courseId = requireArguments().getInt(COURSE_ID_KEY, -1)
        forumId = requireArguments().getInt(FORUM_ID_KEY, -1)
        courseName = requireArguments().getString(COURSE_NAME_KEY, "")

        courseRequestHandler = CourseRequestHandler()
    }

    override fun onStart() {
        requireActivity().title = courseName
        super.onStart()
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?,
    ): View? {
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.swipeRefreshLayout.setOnRefreshListener { refreshContent() }

        val mClickListener = ClickListener { `object`: Any, _: Int ->
            val discussion = `object` as Discussion
            val fragment = DiscussionFragment.newInstance(
                courseId,
                forumId,
                discussion.discussionId,
                courseName
            )
            requireActivity().supportFragmentManager.commit {
                addToBackStack(null)
                replace((requireView().parent as ViewGroup).id, fragment, "ForumDetail")
            }
            true
        }

        mAdapter = Adapter(mClickListener, ArrayList())
        binding.discussions.layoutManager = LinearLayoutManager(requireContext())
        binding.discussions.adapter = mAdapter

        refreshContent()
    }

    private fun refreshContent() {
        binding.swipeRefreshLayout.isRefreshing = true
        CoroutineScope(Dispatchers.IO).launch {
            val realm = Realm.getDefaultInstance()
            val courseDataHandler = CourseDataHandler(realm)

            try {
                val discussions = courseRequestHandler.getForumDicussionsSync(forumId)
                for (discussion in discussions) {
                    discussion.forumId = forumId
                }

                courseDataHandler.setForumDiscussions(forumId, discussions)
                realm.close()

                CoroutineScope(Dispatchers.Main).launch {

                    if (discussions.size == 0) {
                        binding.tvEmpty.text = if (forumId != -1) getString(R.string.no_announcements) else getString(R.string.no_posts_to_display)
                        binding.tvEmpty.visibility = View.VISIBLE
                    } else {
                        mAdapter.clearDiscussions()
                        mAdapter.addDiscussions(discussions)
                    }

                    mAdapter.clearDiscussions()
                    mAdapter.addDiscussions(discussions)
                    binding.swipeRefreshLayout.isRefreshing = false
                }
            } catch (e: IOException) {
                val realmDiscussions = courseDataHandler.getForumDiscussions(forumId)
                val discussions = realm.copyFromRealm(realmDiscussions)
                CoroutineScope(Dispatchers.Main).launch {

                    if (discussions.size == 0) {
                        binding.tvEmpty.text = if (forumId != -1) getString(R.string.no_announcements) else getString(R.string.no_posts_to_display)
                        binding.tvEmpty.visibility = View.VISIBLE
                        Toast
                            .makeText(context, getString(R.string.no_cached_data), Toast.LENGTH_SHORT)
                            .show()
                    } else {
                        Toast.makeText(
                            context, getString(R.string.loading_cached_data),
                            Toast.LENGTH_SHORT,
                        ).show()
                        mAdapter.clearDiscussions()
                        mAdapter.addDiscussions(discussions)
                    }
                    binding.swipeRefreshLayout.isRefreshing = false
                }
            }
        }
    }

    private inner class Adapter(
        private val clickListener: ClickListener,
        val mDiscussions: MutableList<Discussion>,
    ): RecyclerView.Adapter<Adapter.ForumViewHolder>() {
        private val inflater: LayoutInflater = LayoutInflater.from(context)

        fun addDiscussions(discussions: List<Discussion>) {
            mDiscussions.addAll(discussions)
            notifyDataSetChanged()
        }

        fun clearDiscussions() {
            mDiscussions.clear()
            notifyDataSetChanged()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ForumViewHolder {
            val itemBinding = RowForumBinding.inflate(inflater, null, false)
            return ForumViewHolder(itemBinding)
        }

        override fun onBindViewHolder(holder: ForumViewHolder, position: Int) {
            val discussion = mDiscussions[position]
            holder.setIsRecyclable(false)
            holder.bind(discussion)
        }

        override fun getItemCount(): Int {
            return mDiscussions.size
        }

        inner class ForumViewHolder(val itemBinding: RowForumBinding) : RecyclerView.ViewHolder(itemBinding.root) {
            fun bind(discussion: Discussion) {
                Glide.with(itemBinding.userPic.context)
                    .load(Urls.getProfilePicUrl(discussion.userPictureUrl))
                    .into(itemBinding.userPic)

                itemBinding.subject.text = discussion.subject
                itemBinding.userName.text = discussion.userFullName
                itemBinding.message.text = discussion.message
                itemBinding.modifiedTime.text = Utils.formatDate(discussion.timeModified)

                if (!discussion.isPinned) {
                    itemBinding.pinned.visibility = View.GONE
                }
            }

            init {
                itemBinding.root.setOnClickListener {
                    clickListener.onClick(mDiscussions[layoutPosition], layoutPosition)
                }
                itemBinding.clickWrapper.setOnClickListener {
                    val position = layoutPosition
                    clickListener.onClick(mDiscussions[position], position)
                }
            }
        }
    }

    companion object {
        const val COURSE_ID_KEY = "courseId"
        const val FORUM_ID_KEY = "forumId"
        const val COURSE_NAME_KEY = "courseName"

        @JvmOverloads
        fun newInstance(
            courseId: Int = -1,
            forumId: Int = 1,
            courseName: String = "Site News",
        ): ForumFragment {
            val fragment = ForumFragment()
            val args = Bundle()
            args.putInt(FORUM_ID_KEY, forumId)
            args.putInt(COURSE_ID_KEY, courseId)
            args.putString(COURSE_NAME_KEY, courseName)
            fragment.arguments = args
            return fragment
        }
    }
}